Tests/SwiftParserTest/StatementTests.swift

```diff
@@ -783,3 +783,768 @@ final class StatementTests: XCTestCase {
+          name: .identifier("foo")
+        )
+      )
+    )
+  }
+
+  func testThenStmt43() {
+    assertParse(
+      """
+      self.then(0)
+      """,
+      substructure: Syntax(
+        MemberAccessExprSyntax(
+          base: DeclReferenceExprSyntax(baseName: .keyword(.self)),
+          name: .identifier("then")
+        )
+      )
+    )
+  }
+
+  func testThenStmt44() {
+    assertParse(
+      """
+      then /^ then/
+      """,
+      substructure: Syntax(
+        SequenceExprSyntax(
+          elements: .init([
+            DeclReferenceExprSyntax(baseName: .identifier("then")),
+            BinaryOperatorExprSyntax(operator: .binaryOperator("/^")),
+            PostfixOperatorExprSyntax(
+              expression: DeclReferenceExprSyntax(baseName: .identifier("then")),
+              operator: .postfixOperator("/")
+            ),
+          ])
+        )
+      )
+    )
+  }
+
+  func testThenStmt45() {
+    assertParse(
+      """
+      return then
+      """,
+      substructure: Syntax(
+        ReturnStmtSyntax(expression: DeclReferenceExprSyntax(baseName: .identifier("then")))
+      )
+    )
+  }
+
+  func testThenStmt46() {
+    assertParse(
+      """
+      then[0]
+      """,
+      substructure: Syntax(
+        SubscriptCallExprSyntax(
+          calledExpression: DeclReferenceExprSyntax(baseName: .identifier("then")),
+          arguments: .init([
+            .init(expression: IntegerLiteralExprSyntax(0))
+          ])
+        )
+      )
+    )
+  }
+
+  func testThenStmt47() {
+    assertParse(
+      """
+      then: for then in [] {
+        break then
+        continue then
+      }
+      """
+    )
+  }
+
+  func testThenStmt48() {
+    assertParse(
+      """
+      throw then
+      """,
+      substructure: Syntax(
+        ThrowStmtSyntax(expression: DeclReferenceExprSyntax(baseName: .identifier("then")))
+      )
+    )
+  }
+
+  func testThenStmt49() {
+    assertParse(
+      """
+      try then()
+      """
+    )
+  }
+
+  func testThenStmt50() {
+    assertParse(
+      """
+      try then{}
+      """
+    )
+  }
+
+  func testThenStmt51() {
+    assertParse(
+      """
+      try then {}
+      """
+    )
+  }
+
+  func testThenStmt52() {
+    assertParse(
+      """
+      try then + 1
+      """
+    )
+  }
+
+  func testThenStmtDisabled1() {
+    // Make sure it's disabled by default.
+    assertParse(
+      """
+      then1️⃣ 0
+      """,
+      diagnostics: [
+        DiagnosticSpec(
+          message: "consecutive statements on a line must be separated by newline or ';'",
+          fixIts: [
+            "insert newline",
+            "insert ';'",
+          ]
+        )
+      ],
+      fixedSource: """
+        then
+        0
+        """,
+      experimentalFeatures: []
+    )
+  }
+
+  func testThenStmtDisabled2() {
+    // Make sure it's disabled by default.
+    var parser = Parser("then 0")
+    let invalid = StmtSyntax.parse(from: &parser)
+    XCTAssert(invalid.hasError)
```

Comment:
No, this is specifically testing `StmtSyntax.parse` to make sure `then` is disabled by default (since it doesn't check `isStartOfThenStatement`, we need an extra check in `parseThenStatement`)

Reply:
This does actually raise the question of whether `StmtSyntax.parse` itself should check for `atStartOfStatement`, as otherwise it will ignore other heuristics e.g for `yield` statements, and always parse as a statement, even if we would have normally preferred an expr. Or is it expected that it aggressively parses as a statement?

Reply:
> No, this is specifically testing `StmtSyntax.parse` to make sure `then` is disabled by default (since it doesn't check `isStartOfThenStatement`, we need an extra check in `parseThenStatement`)

Ah, good check. Makes sense 👍🏽 



> This does actually raise the question of whether `StmtSyntax.parse` itself should check for `atStartOfStatement`, as otherwise it will ignore other heuristics e.g for `yield` statements, and always parse as a statement, even if we would have normally preferred an expr. Or is it expected that it aggressively parses as a statement?

I think if you’re asking to parse something as a statement, it’s fine if we aggressively parse it as a statement even if that’s not how the parser would do it with context.